### ISSUE TITLE
HLT updates for Run 3: make Scouting producers on-demand

### ISSUE THREAD

**Issue by fwyzard** [cms-sw/cmssw#31186]
The Calo and PF scouting producers are run for every event (barring prescales), independently of the final event selection:
```
process.ScoutingPFOutput = cms.EndPath(
    process.hltGtStage2Digis +
    process.hltPreScoutingPFOutput +
    process.hltFEDSelectorL1 +
    process.HLTPFScoutingPackingSequence +
    process.hltScoutingPrimaryVertexPacker +
    process.hltOutputScoutingPF )

process.ScoutingCaloMuonOutput = cms.EndPath(
    process.hltGtStage2Digis +
    process.hltPreScoutingCaloMuonOutput +
    process.hltFEDSelectorL1 +
    process.hltScoutingCaloPacker +
    process.hltScoutingMuonPackerCalo +
    process.hltScoutingPrimaryVertexPacker +
    process.hltScoutingPrimaryVertexPackerCaloMuon +
    process.hltScoutingTrackPacker +
    process.hltOutputScoutingCaloMuon )

process.PhysicsScoutingMonitorOutput = cms.EndPath(
    process.hltGtStage2Digis +
    process.hltPrePhysicsScoutingMonitorOutput +
    process.hltPrePhysicsScoutingMonitorOutputSmart +
    process.hltScoutingCaloPacker +
    process.hltScoutingMuonPackerCalo +
    process.HLTPFScoutingPackingSequence +
    process.hltScoutingPrimaryVertexPacker +
    process.hltScoutingPrimaryVertexPackerCaloMuon +
    process.hltOutputPhysicsScoutingMonitor )
```

This may result in extra CPU usage - and is especially problematic when some of the product (e.g. pixel tracks) are moved to Tasks.

There are (at least) two solutions to make these producers run only when needed:
  1. make these producers on-demand by putting them in a Task
  2. add a filter (e.g. a smart prescaler) before the producers that selects only the events that will be selected by the output modules

Once the HLT supports Tasks I think Solution 1. would be easier to implement and maintain.

**Comment by fwyzard** [comment]
assign hlt

**Comment by fwyzard** [comment]
@Sam-Harper @gennai @jsalfeld FYI

**Comment by fwyzard** [comment]
A temporary workaround can be something like
```python
from HLTrigger.Configuration.common import *

process.hltPreScoutingCaloMuonOutputSmart = cms.EDFilter( "TriggerResultsFilter",
    l1tIgnoreMaskAndPrescale = cms.bool( False ),
    l1tResults = cms.InputTag( "" ),
    hltResults = cms.InputTag( 'TriggerResults','','@currentProcess' ),
    triggerConditions = process.hltOutputScoutingCaloMuon.SelectEvents.SelectEvents,
    throw = cms.bool( True )
)
insert_modules_after(process, process.hltPreScoutingCaloMuonOutput, process.hltPreScoutingCaloMuonOutputSmart)


process.hltPreScoutingPFOutputSmart = cms.EDFilter( "TriggerResultsFilter",
    l1tIgnoreMaskAndPrescale = cms.bool( False ),
    l1tResults = cms.InputTag( "" ),
    hltResults = cms.InputTag( 'TriggerResults','','@currentProcess' ),
    triggerConditions = process.hltOutputScoutingPF.SelectEvents.SelectEvents,
    throw = cms.bool( True )
)
insert_modules_after(process, process.hltPreScoutingPFOutput, process.hltPreScoutingPFOutputSmart)
```

**Comment by gennai** [comment]
Hi @fwyzard , I need to check the configuration so what I am going to write may be simply wrong, but shouldn't the scouting object packer be run within the scouting paths more than the end path? In this way we would have them run only when needed without the need of further smart-prescales or tasks.
Do I miss anything?

Best,
S.

**Comment by gennai** [comment]
One possible caveat is that in Run2 we had different objects used in different scouting datasets calo vs PF, but regardless the rate at which we will be able to run in Run3 I understood the idea was to unify the reconstruction in only one main path. So running the packer at the end of that path should just work fine.

**Comment by fwyzard** [comment]
Yes, with a single Path (or if all products are produced by all Paths) moving the packers to the end of the Paths would also be fine.

**Comment by Sam-Harper** [comment]
So to summarise our solutions are:

1. packers on tasks and therefore run on demand by construction
2. packers on EndPath but protected by smart prescaler
3. packers on each scouting path in question

2) has the downside where we have to keep that insync with the selection. Andrea's solution fixes this nicely however can this be done in confdb? 
3) has the downside where we need to ensure each path has the proper producers. I think it would be easy to miss (it wouldnt throw an error correct?) and therefore that one path isnt packing objects only gets discovered at the end of the year. 

I agree with andrea that 1) is probably the best, we will push forward with deploying the task updates. 

Best,
Sam

**Comment by fwyzard** [comment]
> 2. ... can this be done in confdb?

Not like this; the smart prescale editor window should help keeping the list of Paths synchronised, but it's easy to forget to do it.

> 3. has the downside where we need to ensure each path has the proper producers.

True... a way to help ensure it is to put all producers in a Sequence, and use that everywhere. Of course it won't prevent all mistakes.

**Comment by fwyzard** [comment]
@acostad FYI

**Comment by missirol** [comment]
+hlt

Since early 2022, the scouting packers are contained in a single sequence (`HLTPixelOnlyPFScoutingPackingSequence`) which runs inside each of the Scouting Paths (not EndPaths). There are currently 3 Paths using this, listed below:
```
DST_Run3_PFScoutingPixelTracking_v
DST_HLTMuon_Run3_PFScoutingPixelTracking_v
MC_Run3_PFScoutingPixelTracking_v
```

**Comment by missirol** [comment]
please close
